docs/proposals/external-lb-source-ip-preservation.md

```diff
@@ -0,0 +1,240 @@
+reprogramming is successful and the API->datapath tables are updated by the cloud provider backend.
+
+## Traffic Steering using Health Checks
+
+This approach requires that all worker nodes in the cluster be programmed into the LB target pool.
+To steer traffic only onto nodes that have endpoints for the service, we program the LB to perform
+node healthchecks. The kube-proxy daemons running on each node will be responsible for responding
+to these healthcheck requests from the cloud provider LB healthchecker.
+kube-proxy already watches for Service and Endpoint changes, it will maintain an in-memory lookup
+table indicating the number of local endpoints for each service.
+For a value of zero, it responds with a health check failure (503 Service Unavailable), and
+success (200 OK) for non-zero values.
+
+Healthchecks are programmable with a min period of 1 second on most cloud provider LBs, and min
+failures to trigger node health state change can be configurable from 2 through 5.
+
+This will allow much faster transition times on the order of 1-5 seconds, and involve no
+API calls to the cloud provider (and hence reduce the impact of API unreliability), keeping the
+time window where traffic might get directed to nodes with no local endpoints to a minimum.
+
+## Choice of traffic steering approaches by individual Cloud Provider implementations
+
+The cloud provider package may choose either of these approaches. kube-proxy will provide these
+healthcheck responder capabilities, regardless of the cloud provider configured on a cluster.
+
+## API Changes
+
+### Local Endpoint Recognition Support
+
+To allow kube-proxy to recognize if an endpoint is local requires that the EndpointAddress struct
+should also contain the NodeName it resides on. This new string field will be read-only and 
+populated *only* by the Endpoints Controller.
+
+### Service Annotation to opt-in for new behaviour
+
+A new annotation `service.alpha.kubernetes.io/node-local-loadbalancer` will be recognized
+by the service controller. Services that wish to opt-in to the new LoadBalancer behaviour must
+annotate the Service to request the new ESIPP behaviour as documented in the section below.
+
+### NodePort allocation for HealthChecks
+
+An additional nodePort allocation will be necessary for services that are of type LoadBalancer and
+have the new annotation specified. This additional nodePort is necessary for kube-proxy to listen for
+healthcheck requests on all nodes.
+This NodePort will be added as an annotation to the Service after allocation (in the alpha release).
+
+## Behavior Changes expected
+
+### External Traffic Blackholed on nodes with no local endpoints
+
+When the last endpoint on the node has gone away and the LB has not marked the node as unhealthy,
+worst-case window size = (N+1) * HCP, where N = minimum failed healthchecks and HCP = Health Check Period,
+external traffic will still be steered to the node. This traffic will be blackholed and not forwarded
```

Comment:
I can buy this, but we need to be VERY clear that this is a departure from
existing nodeports.  Or we need another thing that is like nodeports but
with the new semantic.

On Mon, Aug 15, 2016 at 10:52 AM, Prashanth B <notifications@github.com>
wrote:

> In docs/proposals/external-lb-source-ip-preservation.md
> <https://github.com/kubernetes/kubernetes/pull/30105#discussion_r74802541>
> :
>
> > +annotate the Service to request the new ESIPP behaviour as documented in the section below.
> > +
> > +### NodePort allocation for HealthChecks
> > +
> > +An additional nodePort allocation will be necessary for services that are of type LoadBalancer and
> > +have the new annotation specified. This additional nodePort is necessary for kube-proxy to listen for
> > +healthcheck requests on all nodes.
> > +This NodePort will be added as an annotation to the Service after allocation (in the alpha release).
> > +
> > +## Behavior Changes expected
> > +
> > +### External Traffic Blackholed on nodes with no local endpoints
> > +
> > +When the last endpoint on the node has gone away and the LB has not marked the node as unhealthy,
> > +worst-case window size = (N+1) * HCP, where N = minimum failed healthchecks and HCP = Health Check Period,
> > +external traffic will still be steered to the node. This traffic will be blackholed and not forwarded
>
> During that window, users will experience a real outage. Is that better
> than getting the wrong source IP?
>
> I'd say yes for the same reason discussed below, it's easier to engineer
> around blackholing either with client retries or by just labeling nodes
> with role=lb and loading it up with replicas. If a client ip is
> blacklisted it shoudn't be allowed through, no matter what, IMO.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30105/files/ac8a27036cf9811c9de760c578958948370629f1#r74802541>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVN_ncL-7QkYA5CQN5XroCJyDxBv-ks5qgKdWgaJpZM4JdJyL>
> .
>